File: src/ir/glyphIR.ts
import type {
  CallArgAst,
  ContextAst,
  ExpressionAst,
  ObjectEntryAst,
  ProgramAst,
  StatementAst,
  ToolCallAst
} from "../language/ast.js";
import { parseGlyph } from "../language/parser.js";

export const GLYPH_IR_VERSION = "0.1" as const;

export type GlyphPrimitiveValue = string | number | boolean | null;

export type GlyphIRValue =
  | GlyphPrimitiveValue
  | GlyphIRValue[]
  | { [key: string]: GlyphIRValue }
  | { var: string }
  | { ctx: string };

export interface GlyphIR {
  version: typeof GLYPH_IR_VERSION;
  goal?: string;
  context: Record<string, GlyphPrimitiveValue | GlyphIRValue[] | { [key: string]: GlyphIRValue }>;
  flows: GlyphIRFlow[];
}

export interface GlyphIRFlow {
  name: string;
  steps: GlyphIRStep[];
}

export type GlyphIRStep = GlyphToolStep | GlyphRepairStep;

export interface GlyphToolStep {
  kind: "tool";
  id: string;
  op: string;
  args: Record<string, GlyphIRValue>;
  assignTo?: string;
}

export interface GlyphRepairStep {
  kind: "repair";
  id: string;
  targetVar: string;
  reportVar: string;
  maxIterations: number;
  steps: GlyphIRStep[];
}

const POSITIONAL_ARG_NAMES: Record<string, string[]> = {
  SPEC: ["input"],
  PLAN: ["input"],
  GEN: ["input"],
  CHECK: ["target"],
  FIX: ["target", "report"],
  PATCH: ["target", "instructions"],
  SUM: ["target"],
  SUMMARIZE: ["target"],
  ASK: ["question", "options"],
  EXPORT: ["target", "format"],
  RUN: ["command", "target"],
  READ: ["path"],
  WRITE: ["path", "content"]
};

export function parseGlyphToIR(source: string): GlyphIR {
  return compileAstToIR(parseGlyph(source));
}

export function compileAstToIR(ast: ProgramAst): GlyphIR {
  let stepCounter = 0;
  const nextStepId = () => `step_${++stepCounter}`;

  const compileStep = (step: StatementAst): GlyphIRStep => {
    if (step.kind === "ToolCall") {
      return compileToolStep(step, nextStepId());
    }

    return {
      kind: "repair",
      id: nextStepId(),
      targetVar: step.target,
      reportVar: step.report,
      maxIterations: step.max,
      steps: step.steps.map(compileStep)
    };
  };

  return {
    version: GLYPH_IR_VERSION,
    goal: ast.goal,
    context: ast.context ? compileContext(ast.context) : {},
    flows: ast.flows.map((flow) => ({
      name: flow.name,
      steps: flow.steps.map(compileStep)
    }))
  };
}

function compileToolStep(step: ToolCallAst, id: string): GlyphToolStep {
  return {
    kind: "tool",
    id,
    op: step.op.toUpperCase(),
    args: compileCallArgs(step.op.toUpperCase(), step.args),
    assignTo: step.assignTo
  };
}

function compileCallArgs(op: string, args: CallArgAst[]): Record<string, GlyphIRValue> {
  const record: Record<string, GlyphIRValue> = {};
  const positionalNames = POSITIONAL_ARG_NAMES[op] ?? ["input"];
  let positionalIndex = 0;

  for (const arg of args) {
    const name = arg.name ?? positionalNames[positionalIndex] ?? `arg${positionalIndex + 1}`;
    positionalIndex += arg.name ? 0 : 1;

    if (record[name] !== undefined) {
      throw new Error(`Duplicate argument "${name}" for ${op}`);
    }

    record[name] = expressionToIRValue(arg.value);
  }

  return record;
}

function compileContext(context: ContextAst): GlyphIR["context"] {
  return Object.fromEntries(context.entries.map((entry) => [entry.key, expressionToJsonLiteral(entry.value)]));
}

function expressionToJsonLiteral(expression: ExpressionAst): GlyphIR["context"][string] {
  switch (expression.kind) {
    case "StringLiteral":
    case "NumberLiteral":
    case "BooleanLiteral":
      return expression.value;
    case "ArrayLiteral":
      return expression.items.map(expressionToJsonLiteral);
    case "ObjectLiteral":
      return Object.fromEntries(expression.entries.map((entry) => [entry.key, expressionToJsonLiteral(entry.value)]));
    case "VarRef":
    case "CtxRef":
      throw new Error("Context declarations must use literal JSON-compatible values");
  }
}

function expressionToIRValue(expression: ExpressionAst): GlyphIRValue {
  switch (expression.kind) {
    case "StringLiteral":
    case "NumberLiteral":
    case "BooleanLiteral":
      return expression.value;
    case "ArrayLiteral":
      return expression.items.map(expressionToIRValue);
    case "ObjectLiteral":
      return objectEntriesToIR(expression.entries);
    case "VarRef":
      return { var: expression.name };
    case "CtxRef":
      return { ctx: expression.path.join(".") };
  }
}

function objectEntriesToIR(entries: ObjectEntryAst[]): Record<string, GlyphIRValue> {
  const result: Record<string, GlyphIRValue> = {};

  for (const entry of entries) {
    if (result[entry.key] !== undefined) {
      throw new Error(`Duplicate object key "${entry.key}"`);
    }
    result[entry.key] = expressionToIRValue(entry.value);
  }

  return result;
}

File: src/ir/glyph_ir.rs
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value, json};

use crate::language::ast::{
    CallArgAst, ContextAst, ExpressionAst, ObjectEntryAst, ProgramAst, StatementAst, ToolCallAst,
};
use crate::language::errors::GlyphSyntaxError;
use crate::language::parser::parse_glyph;

pub const GLYPH_IR_VERSION: &str = "0.1";

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GlyphIr {
    pub version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub goal: Option<String>,
    pub context: Map<String, Value>,
    pub flows: Vec<GlyphIrFlow>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GlyphIrFlow {
    pub name: String,
    pub steps: Vec<GlyphIrStep>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum GlyphIrStep {
    #[serde(rename = "tool")]
    Tool(GlyphToolStep),
    #[serde(rename = "repair")]
    Repair(GlyphRepairStep),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GlyphToolStep {
    pub id: String,
    pub op: String,
    pub args: Map<String, Value>,
    #[serde(rename = "assignTo", skip_serializing_if = "Option::is_none")]
    pub assign_to: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GlyphRepairStep {
    pub id: String,
    #[serde(rename = "targetVar")]
    pub target_var: String,
    #[serde(rename = "reportVar")]
    pub report_var: String,
    #[serde(rename = "maxIterations")]
    pub max_iterations: usize,
    pub steps: Vec<GlyphIrStep>,
}

pub fn parse_glyph_to_ir(source: &str) -> Result<GlyphIr, GlyphSyntaxError> {
    Ok(compile_ast_to_ir(&parse_glyph(source)?))
}

pub fn compile_ast_to_ir(ast: &ProgramAst) -> GlyphIr {
    let mut counter = 0usize;
    let mut next_id = || {
        counter += 1;
        format!("step_{counter}")
    };

    GlyphIr {
        version: GLYPH_IR_VERSION.to_string(),
        goal: ast.goal.clone(),
        context: ast
            .context
            .as_ref()
            .map(compile_context)
            .unwrap_or_default(),
        flows: ast
            .flows
            .iter()
            .map(|flow| GlyphIrFlow {
                name: flow.name.clone(),
                steps: flow
                    .steps
                    .iter()
                    .map(|step| compile_step(step, &mut next_id))
                    .collect(),
            })
            .collect(),
    }
}

fn compile_step(step: &StatementAst, next_id: &mut impl FnMut() -> String) -> GlyphIrStep {
    match step {
        StatementAst::ToolCall(call) => GlyphIrStep::Tool(compile_tool_step(call, next_id())),
        StatementAst::RepairBlock(block) => GlyphIrStep::Repair(GlyphRepairStep {
            id: next_id(),
            target_var: block.target.clone(),
            report_var: block.report.clone(),
            max_iterations: block.max,
            steps: block
                .steps
                .iter()
                .map(|inner| compile_step(inner, next_id))
                .collect(),
        }),
    }
}

fn compile_tool_step(step: &ToolCallAst, id: String) -> GlyphToolStep {
    let op = step.op.to_uppercase();
    GlyphToolStep {
        id,
        op: op.clone(),
        args: compile_call_args(&op, &step.args),
        assign_to: step.assign_to.clone(),
    }
}

fn compile_call_args(op: &str, args: &[CallArgAst]) -> Map<String, Value> {
    let mut record = Map::new();
    let positional_names = positional_arg_names(op);
    let mut positional_index = 0usize;

    for arg in args {
        let name = match &arg.name {
            Some(name) => name.clone(),
            None => {
                let name = positional_names
                    .get(positional_index)
                    .copied()
                    .unwrap_or("input")
                    .to_string();
                positional_index += 1;
                name
            }
        };

        record.insert(name, expression_to_ir_value(&arg.value));
    }

    record
}

fn compile_context(context: &ContextAst) -> Map<String, Value> {
    context
        .entries
        .iter()
        .map(|entry| (entry.key.clone(), expression_to_json_literal(&entry.value)))
        .collect()
}

fn expression_to_json_literal(expression: &ExpressionAst) -> Value {
    match expression {
        ExpressionAst::String(value) => Value::String(value.clone()),
        ExpressionAst::Number(value) => Value::Number(value.clone()),
        ExpressionAst::Boolean(value) => Value::Bool(*value),
        ExpressionAst::Array(items) => {
            Value::Array(items.iter().map(expression_to_json_literal).collect())
        }
        ExpressionAst::Object(entries) => object_entries_to_json(entries),
        ExpressionAst::VarRef(_) | ExpressionAst::CtxRef(_) => {
            panic!("context declarations must use literal JSON-compatible values")
        }
    }
}

fn expression_to_ir_value(expression: &ExpressionAst) -> Value {
    match expression {
        ExpressionAst::String(value) => Value::String(value.clone()),
        ExpressionAst::Number(value) => Value::Number(value.clone()),
        ExpressionAst::Boolean(value) => Value::Bool(*value),
        ExpressionAst::Array(items) => {
            Value::Array(items.iter().map(expression_to_ir_value).collect())
        }
        ExpressionAst::Object(entries) => object_entries_to_ir(entries),
        ExpressionAst::VarRef(name) => json!({ "var": name }),
        ExpressionAst::CtxRef(path) => json!({ "ctx": path.join(".") }),
    }
}

fn object_entries_to_json(entries: &[ObjectEntryAst]) -> Value {
    Value::Object(
        entries
            .iter()
            .map(|entry| (entry.key.clone(), expression_to_json_literal(&entry.value)))
            .collect(),
    )
}

fn object_entries_to_ir(entries: &[ObjectEntryAst]) -> Value {
    Value::Object(
        entries
            .iter()
            .map(|entry| (entry.key.clone(), expression_to_ir_value(&entry.value)))
            .collect(),
    )
}

fn positional_arg_names(op: &str) -> &'static [&'static str] {
    match op {
        "SPEC" => &["input"],
        "PLAN" => &["input"],
        "GEN" => &["input"],
        "CHECK" => &["target"],
        "FIX" => &["target", "report"],
        "PATCH" => &["target", "instructions"],
        "SUM" | "SUMMARIZE" => &["target"],
        "ASK" => &["question", "options"],
        "EXPORT" => &["target", "format"],
        "RUN" => &["command", "target"],
        "READ" => &["path"],
        "WRITE" => &["path", "content"],
        _ => &["input"],
    }
}
